### PR TITLE
remove speech-rule-engine integration

### LINKED ISSUE #207: move SRE integration out of mj-node
As per F2F (cf. #191), the SRE integration will be moved out of mathjax-node; there will be alternative modules.

### PR DESCRIPTION
**Pull request by pkra**
Resolves #207

### CODE REVIEW

**Comment by pkra** [comment]
Ready for review.

**Comment by pkra** [comment]
@dpvc @zorkow I did a [poll on twitter](https://twitter.com/pkrautz/status/793752948014710784) (some people responded in tweets, but those were all against "prefixes"); not exactly scientific but I think we should drop the equation title. What do you think?

(FWIW I have been thinking about proposing a redefinition of role=math, and prefixes would be one item that AT could provide an opt-in for.)

**Comment by pkra** [comment]
Also, https://github.com/pkra/mathjax-node-sre is now actually working.

**Comment by pkra** [comment]
> I think we should drop the equation title. What do you think?

I've updated the commit.

File: lib/main.js [hunk 1]
```diff
@@ -549,30 +540,18 @@ function GetSpeech(result) {
...
   if (!data.html && !data.htmlNode) return;
   var jax = MathJax.Hub.getAllJax()[0]; if (!jax) return;
   var script = jax.SourceElement(), html = script.previousSibling;
 
-  if (data.speakText) {
-    var label = html.querySelector("[aria-label]");
-    if (label) label.removeAttribute("aria-label");
+  // add dummy speech text if there isn't one
+  for (var i=0, m=html.childNodes.length; i < m; i++)
+    html.childNodes[i].setAttribute("aria-hidden",true);
```
Comment (zorkow) [inline lib/main.js]: See remark below.

File: lib/main.js [hunk 2]
```diff
@@ -608,17 +587,16 @@ function GetSVG(result) {
...
-    svg.setAttribute("aria-labelledby",id+"-Title "+id+"-Desc");
-    for (var i=0, m=svg.childNodes.length; i < m; i++)
-      svg.childNodes[i].setAttribute("aria-hidden",true);
-    var node = MathJax.HTML.Element("desc",{id:id+"-Desc"},[result.speakText]);
-    svg.insertBefore(node,svg.firstChild);
-    node = MathJax.HTML.Element("title",{id:id+"-Title"},["Equation"]);
-    svg.insertBefore(node,svg.firstChild);
-  }
+  for (var i=0, m=svg.childNodes.length; i < m; i++)
+    svg.childNodes[i].setAttribute("aria-hidden",true);
```
Comment (zorkow) [inline lib/main.js]: In future coding style this should become a utility function (e.g., ariaHideChildren(node)), together as in lines 551/552 above.

File: lib/main.js [hunk 3]
```diff
@@ -608,17 +587,16 @@ function GetSVG(result) {
...
+  // Note: if aria-label exists, getSpeech preserved it in speakText
+  // remove aria-label since labelled-by desc is preferred
+  svg.removeAttribute("aria-label");
```
Comment (zorkow) [inline lib/main.js]: Do we just throw away the original aria label, if there is one?

File: test/issue207.js [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    var element = document.querySelector('.MJXc-display');
+    var result = element.getAttribute('aria-label');
+    t.equal(result, input1, 'HTML output contains dummy speechText');
+    var document = jsdom(data.mml).defaultView.document;
+    var element = document.querySelector('math');
+    var result = element.getAttribute('alttext');
+    t.equal(result, input1, 'MathML output contains dummy speechText');
+    var document = jsdom(data.svg).defaultView.document;
+    var desc = document.querySelector('desc');
+    t.equal(result, desc.innerHTML, 'SVG output contains dummy speechText');
```
Comment (zorkow) [inline test/issue207.js]: Refactor into function with code in the second unit test.

File: test/issue207.js [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    var result = element.getAttribute('aria-label');
+    t.equal(result, expected, 'HTML output from MathML contains original speechText');
+    var document = jsdom(data.mml).defaultView.document;
+    var element = document.querySelector('math');
+    var result = element.getAttribute('alttext');
+    t.equal(result, expected, 'MathML output from MathML contains original dummy speechText');
+    var document = jsdom(data.svg).defaultView.document;
+    var desc = document.querySelector('desc');
+    t.equal(expected, desc.innerHTML, 'SVG output from MathML contains original dummy speechText');
+  });
```
Comment (zorkow) [inline test/issue207.js]: Refactor, see remark above.

File: test/issue207.js [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    mml: true
+  }, function(data) {
+    var document = jsdom(data.html).defaultView.document;
+    var element = document.querySelector('.MJXc-display');
+    var result = element.getAttribute('aria-label');
+    t.equal(result, expected, 'HTML output from MathML contains original speechText');
+    var document = jsdom(data.mml).defaultView.document;
+    var element = document.querySelector('math');
+    var result = element.getAttribute('alttext');
+    t.equal(result, expected, 'MathML output from MathML contains original dummy speechText');
```
Comment (zorkow) [inline test/issue207.js]: Call parameters actual and expected. 
(Convention dictacted that expected comes before actual, but that's just silly...)

File: lib/main.js [hunk 7]
```diff
@@ -532,15 +523,15 @@ function GetMML(result) {
...
 function GetSpeech(result) {
-  if (!data.speakText) return;
-  speech.setupEngine({semantics: true, domain: data.speakRuleset, style: data.speakStyle});
-  result.speakText = speech.toSpeech(result.mml);
+  result.speakText = "Equation";
+  if (data.format !== "MathML") result.speakText = data.math;
   if (!data.mml) return;
   var jax = MathJax.Hub.getAllJax()[0];
-  jax.root.alttext = result.speakText;
+  if (jax.root.alttext) result.speakText = jax.root.alttext;
```
Comment (pkra) [inline lib/main.js]: I don't understand. The next line sets the `alttext` if it didn't exist.

Comment (pkra) [inline lib/main.js]: Since I'm not sure how the reviews are pushed out in notifications: I just drastically edited my earlier comment.

File: lib/main.js (hunk 3 above)
Comment (pkra) [inline lib/main.js]: Yes because your recommendation had been to use `desc` instead of an aria-label for SVGs.

The aria-label occurs only via core mathjax: if MathML input has an `alttext`, core mathjax adds that as an aria-label. But in that case we used the `alttext` value for `speakText` above and thus the `desc` will be the same as the label. Moving the change upstream would eliminate this part.

File: test/issue207.js (hunk 4 above)
Comment (pkra) [inline test/issue207.js]: Hm. I don't see how to refactor this without making the test worse. I don't want to give up on having differentiating messages from the tests. I also don't want to create some complicated function that generates fitting tests and messages dynamically because I worry that this will make the test less readable. 

Do you have any suggestions?

File: test/issue207.js (hunk 6 above)
Comment (pkra) [inline test/issue207.js]: > Call parameters actual and expected.

OK.

File: test/issue207.js (hunk 5 above)
Comment (pkra) [inline test/issue207.js]: See above.

File: lib/main.js (hunk 7 above)
Comment (zorkow) [inline lib/main.js]: What I meant was if alttext is "" rather than non-existent, then it will be overwritten by speakText. Could the case happen that alttext is ever deliberately left empty? If yes, then the test should be refined to check for undefinedness, if no, it's fine.
[This is just a sanity-check. Sometimes careless treatment of semi-boolean values, i.e., "", '', 0, can lead to unintended effects and errors. I ran often enough into that problem, e.g,. when iterating over lists of integers.]

File: lib/main.js (hunk 3 above)
Comment (zorkow) [inline lib/main.js]: sgtm.

File: lib/main.js (hunk 7 above)
Comment (pkra) [inline lib/main.js]: Thanks for clarifying! I had misunderstood and had not thought of this imprecision. I think overwriting falsy values is ok but I'm happy to change it.

File: test/issue207.js (hunk 5 above)
Comment (zorkow) [inline test/issue207.js]: In addition you could refactor the actual calls to typeset into a single function, so the actual calls to the test function will be of the form:
`callMjSpeechTest(format, math, expected);`
e.g., `callMjSpeechTest('TeX', 'x', 'x');`
which will make the code and the principle of the tests a lot more readable.

Comment (pkra) [inline test/issue207.js]: Thanks for clarifying. As I wrote earlier, the problem I saw with this kind of refactor is that the tap messages are losing important information. But I have an idea for working around that.

**Comment by zorkow** [comment]
Certainly weird. It seems to duplicate comments.

**Comment by pkra** [comment]
I mostly dislike that it opens a new review when I'm just trying to respond to your comments. Especially since I'd have assumed that I can't review my own PR.

**Comment by zorkow** [comment]
Well, I sometimes review my own PRs in SRE, but remember github complaining, that I can't open a review on my own PR. But apparently you can review someone's review on your PR. Is that a meta-review???

Anyway, I would have thought that it clearly groups replies chronologically, if that makes sense.

**Comment by pkra** [comment]
@zorkow I've refactored the test.

File: lib/main.js [hunk 8]
```diff
@@ -60,14 +59,6 @@ var defaults = {
   mmlNode: false,              // DOM element for MML output?
   svg: false,                     // return svg output?
   svgNode: false,              // DOM element for SVG output?
-
-  speakText: false,               // add spoken annotations to svg output?
```
Comment (dpvc) [inline lib/main.js]: Your current implementation always returns `speakText` in the `results` object, and always adds some speech text to the SVG, HTML, or MML output.  I'd prefer to retain the option of adding speech text (though it could be true by default), and only add it and return it when requested.

File: lib/main.js [hunk 9]
```diff
@@ -532,15 +523,15 @@ function GetMML(result) {
...
 //
-//  Creates the speech string and updates the MathML to include it, if needed
+//  Creates dummy speech string and updates the MathML to include it, if needed
 //
 function GetSpeech(result) {
-  if (!data.speakText) return;
-  speech.setupEngine({semantics: true, domain: data.speakRuleset, style: data.speakStyle});
-  result.speakText = speech.toSpeech(result.mml);
+  result.speakText = "Equation";
+  if (data.format !== "MathML") result.speakText = data.math;
```
Comment (dpvc) [inline lib/main.js]: Note that it is possible for this to make the `speakText` be blank (if `data.math` is blank, e.g., from `\(\)` and input is TeX).

File: lib/main.js [hunk 10]
```diff
@@ -608,17 +587,16 @@ function GetSVG(result) {
...
-    svg.insertBefore(node,svg.firstChild);
-    node = MathJax.HTML.Element("title",{id:id+"-Title"},["Equation"]);
-    svg.insertBefore(node,svg.firstChild);
-  }
+  for (var i=0, m=svg.childNodes.length; i < m; i++)
+    svg.childNodes[i].setAttribute("aria-hidden",true);
+  // Note: if aria-label exists, getSpeech preserved it in speakText
+  // remove aria-label since labelled-by desc is preferred
+  svg.removeAttribute("aria-label");
+  ID++; var id = "MathJax-SVG-"+ID;
```
Comment (dpvc) [inline lib/main.js]: Why not just add `'-Desc'` to the id at this point rather than adding it twice below (now that you no longer need `id+'-Title'`)?

**Comment by dpvc** [comment]
Here are a few other comments that I couldn't add as line comments, since they are not about lines in the change file.

* The blank line at [600](https://github.com/mathjax/MathJax-node/blob/444451543ab51b76f23c709f630d3c355c6ea6a6/lib/main.js#L600) is not needed.  (There are some other spacing issues that Volker may complain about as well...)

* [Lines 740 to 742](https://github.com/mathjax/MathJax-node/blob/444451543ab51b76f23c709f630d3c355c6ea6a6/lib/main.js#740-742) are  no longer needed, since the `results.mml` is only created when actually needed (it used to be created for speech creation even when not requested by `data.mml`, and had to be cleaned up here).  If you end up going with optional speech string creation, as I requested above, then you might use similar code to this to remove `results.speakText` when not needed.

* The speech options need to be removed from the command-line tools.

I looked over the file as a whole, and found some issues that don't come directly from this pull request, but should be handled at some point anyway:

* At [line 606](https://github.com/mathjax/MathJax-node/blob/444451543ab51b76f23c709f630d3c355c6ea6a6/lib/main.js#L606-L613), you no longer need a separate variable (it used to be used to be needed for writing the external files, but that was removed earlier).

* At [line 573](https://github.com/mathjax/MathJax-node/blob/444451543ab51b76f23c709f630d3c355c6ea6a6/lib/main.js#L573) and [line 615](https://github.com/mathjax/MathJax-node/blob/444451543ab51b76f23c709f630d3c355c6ea6a6/lib/main.js#L615), the nodes being returned should be removed from the DOM before being sent back in the `results` object (otherwise the caller then has access to the internal jsdom document).  Alternatively, they could be cloned rather than disrupting the DOM.

* At [line 718](https://github.com/mathjax/MathJax-node/blob/444451543ab51b76f23c709f630d3c355c6ea6a6/lib/main.js#L718), `MathJax.OutputJax.CommonHTML` can be replaced by `CHTML`, since that is defined earlier.

* Similarly, at [line 750](https://github.com/mathjax/MathJax-node/blob/444451543ab51b76f23c709f630d3c355c6ea6a6/lib/main.js#L750), `MathJax.ElementJax.mml` can be replaced by `MML`.

* In [lines 616 through 618](https://github.com/mathjax/MathJax-node/blob/444451543ab51b76f23c709f630d3c355c6ea6a6/lib/main.js#L616-618), should these be `svgWidth`, `svgHeight`, and `svgStyles`, since they are specific to the SVG output?

These are what I see at this point.  Hope that helps.

File: lib/main.js (hunk 8 above)
Comment (pkra) [inline lib/main.js]: OK

File: lib/main.js (hunk 9 above)
Comment (pkra) [inline lib/main.js]: Right. Cf. also Volker's comment below.

File: lib/main.js (hunk 10 above)
Comment (pkra) [inline lib/main.js]: No reason. Just copy and paste from the original (where a title was also present and thus you had split this up).

**Comment by pkra** [comment]
> The speech options need to be removed from the command-line tools.

Since you suggested the switch should stay, they should stay as well.

> I looked over the file as a whole, and found some issues that don't come directly from this pull request, but should be handled at some point anyway:

I suggest we file those separately.

**Comment by pkra** [comment]
I hope I've addressed all review comments.

**Comment by pkra** [comment]
Sorry. I broke the tests (again).

**Comment by pkra** [comment]
Fixed the tests.

File: lib/main.js (hunk 8 above)
Comment (dpvc) [inline lib/main.js]: Thanks.  Note, however, that the comment says it adds to svg, but it really adds to all output formats.

**Comment by dpvc** [comment]
> >The speech options need to be removed from the command-line tools.
>
> Since you suggested the switch should stay, they should stay as well.

There were several options to control SRE that are no longer needed and should be removed from them.  Also, now that the default for `speakText` is `true`, you may want the command line tools to default that way as well, which they currently don't.

**Comment by dpvc** [comment]
Also, you will want to remove the `*2stree` command-line tools, since those functions have been removed.

File: lib/main.js [hunk 11]
```diff
@@ -608,17 +591,17 @@ function GetSVG(result) {
...
-    node = MathJax.HTML.Element("title",{id:id+"-Title"},["Equation"]);
-    svg.insertBefore(node,svg.firstChild);
+    // Note: if aria-label exists, getSpeech preserved it in speakText
+    // remove aria-label since labelled-by desc is preferred
+    svg.removeAttribute("aria-label");
   }
+  ID++; var id = "MathJax-SVG-"+ID+"-Desc";
+  svg.setAttribute("aria-labelledby",id);
+  var node = MathJax.HTML.Element("desc",{id:id},[result.speakText]);
+  svg.insertBefore(node,svg.firstChild);
```
Comment (dpvc) [inline lib/main.js]: Shouldn't these four lines (601-604) be inside the test for `data.speakText` as well?

Comment (pkra) [inline lib/main.js]: Right.

**Comment by pkra** [comment]
> There were several options to control SRE that are no longer needed and should be removed from them. 

Right. But cf #208, I still think they should all live in an entirely separate package anyway.

**Comment by pkra** [comment]
> Also, now that the default for speakText is true, you may want the command line tools to default that way as well, which they currently don't.

Looks like they do?

**Comment by pkra** [comment]
PTAL.

**Comment by pkra** [comment]
> Looks like they do?

Ah, I was wrong about that. Confusing the type with the default. Will add a default.

**Comment by pkra** [comment]
PTAL.

**Comment by pkra** [comment]
I've been doing [a bit of testing today](https://s.codepen.io/pkra/debug/RodYOJ) and realized that we should change the SVG approach: a `title` tag has a better chance of actually producing a result on AT.

**Comment by pkra** [comment]
PTAL.

**Comment by dpvc** [comment]
I'm OK with it at this point.  @zorkow needs to accept your changes to complete his review.  Then we can merge.

File: test/issue207.js (hunk 6 above)
Comment (zorkow) [inline test/issue207.js]: Don't redeclare variables in the same scope. Otherwise, the interpreted does a lot of hoisting, which is unnecessary.

**Comment by zorkow** [comment]
Get rid of the superfluous "var" declarations. Then it's OK.

**Comment by pkra** [comment]
> Get rid of the superfluous "var" declarations. Then it's OK.

Uhm. Which commit were you seeing these on? I have trouble spotting them :-(

**Comment by pkra** [comment]
> Get rid of the superfluous "var" declarations. Then it's OK.

Cleared up F2F.
